### DISCUSSION TITLE
Filtering aeroway for airports with IATA code.

### DISCUSSION THREAD

**Discussion by FunMiles** [systemed/tilemaker#784] [Q&A] [answered]
First, let me say thank you for all the great work. I have been looking at `tilemaker` for quite a while and I am happy that I can now process the whole OSM planet file on my MacbookPro M3 with 128GB in about 23 minutes, either in RAM or using the internal SSD for temporary storage. Amazing work!

Now for my quest:
What would be the best way to filter all _aeroway_ nodes to only keep airports with a non empty _iata_ code, e.g.?
For airport labels, the _iata_ attribute is present and can be used to answer some questions. But it seems that such attribute (or icao) are not present on nodes of type _runway_, _taxiway_, etc...

Is there a way to retrieve any _aerodrome_ node within a given distance of a node such as a _runway_ or _taxiway_?

One fallback idea, would be to create a GeoJson or shape file with a square around every center of each airport with an IATA code (I have the latitude, longitude of pretty much every IATA airport in a separate file). With such a file, I could use similar enquiries to the ones in the examples to filter by country.
Can tilemaker be used to output such a file automatically, such as using Lua code to write a shapefile or GeoJson with a box every airport label found in the OSM file? Does the Lua code run on a single thread or is there one Lua instance per thread? In the latter case, writing such a file would be a lot more difficult.
Finally, does _FindIntersecting_ work on a layer of any type, including those from a GeoJson file?

PS: I was reading the docs again and found something that might help:
`init_function`: First it seems to make clear that there must be more than one Lua instance. Second, I could probably read the list of IATA airports in this function and provide a query function to find out if a node is within a given distance of any of the IATA airports.
Does that sound reasonable?

**Comment by cldellow** [comment, marked as answer]
> Can tilemaker be used to output such a file automatically, such as using Lua code

I'm excited to see more people treating tilemaker like a general-purpose OSM processing tool. :) I do something similar for my hobby project (https://www.hikeratlas.com/blog/)

> Does the Lua code run on a single thread or is there one Lua instance per thread?

One instance per thread. The instance is created at the start of each phase (read nodes, read ways, read relations) and destroyed at the end.

> does FindIntersecting work on a layer of any type, including those from a GeoJson file?

Yes

> [if there are multiple Lua instances] writing such a file would be a lot more difficult.

Yes, you'll get corrupt output unless you take precautions. You can use `flock` to synchronize access to a shared file. I wrote a short blog post about it here: https://www.hikeratlas.com/blog/lua-atomic-file-append

> Can tilemaker be used to output such a file automatically, such as using Lua code to write a shapefile or GeoJson with a box every airport label found in the OSM file?

Yes, but there are a couple tricks needed.

If you want to write JSON, you'll need a JSON library. [json.lua](https://github.com/rxi/json.lua) is a pure Lua JSON library that I've used successfully.

If you want tilemaker to output GeoJSON geometries, you'll need to patch it. [This patch](https://github.com/cldellow/tilemaker/pull/4/commits/cb392234550692155e32114b80a2e9c50dd57239) adds a `GeoJSON(bool area)` function that you can call from a `node_function`, `way_function` or `relation_function` to get the corresponding GeoJSON geometry (...which you can then write out as JSON using json.lua + flock).

But for your specific problem it sounds like you might be able to get by without full geometries? e.g. you could use the `Centroid()` function to get a lat/lng, and then handcraft a GeoJSON geometry that describes a bounding box around that lat/lng with some buffer. Then that GeoJSON file could be used as input to a future run.

> I could probably read the list of IATA airports in this function and provide a query function to find out if a node is within a given distance of any of the IATA airports.

Yup, that'd work. If your list of data is large, the overhead of reading it in once per thread and once per phase might be annoying. Probably that's not an issue for something like a list of airports, but if it is, you might consider storing it in a sqlite DB. `lua-sql-sqlite3` works, and is available in the tilemaker Docker image since https://github.com/systemed/tilemaker/commit/7c838e094320073b596d365496071d00992d37fa

**Reply by FunMiles** [reply]
> I'm excited to see more people treating tilemaker like a general-purpose OSM processing tool. :) I do something similar for my hobby project (https://www.hikeratlas.com/blog/)

I had a look at your site. Interesting stuff.

> But for your specific problem it sounds like you might be able to get by without full geometries? e.g. you could use the `Centroid()` function to get a lat/lng, and then handcraft a GeoJSON geometry that describes a bounding box around that lat/lng with some buffer. Then that GeoJSON file could be used as input to a future run.
> [ ... ]
> > I could probably read the list of IATA airports in this function and provide a query function to find out if a node is within a given distance of any of the IATA airports.
> 
> Yup, that'd work. If your list of data is large, the overhead of reading it in once per thread and once per phase might be annoying. Probably that's not an issue for something like a list of airports, but if it is, you might consider storing it in a sqlite DB. `lua-sql-sqlite3` works, and is available in the tilemaker Docker image since [7c838e0](https://github.com/systemed/tilemaker/commit/7c838e094320073b596d365496071d00992d37fa)

I am going to go that way. Use the `Centroid` of the _way/node_ and search for IATA airport coordinates within a range around the centroid's longitude/latitude. Whether I read the file and do comparison by hand or use sqlite, I have not decided yet. Reading the file is the most straightforward of the two. For doing a search on a sqlite database, I'd need to add something like an R-Tree index (https://www.sqlite.org/rtree.html) to the table of airports and handle the added query complexity. Perhaps for now, it would be easier to do a loop over the entries. I can easily sort the airport entries by longitude and reduce the number of entries to consider for each individual query to something quite small.
